### PR TITLE
2510 AI Brief input logic improvements + analytics

### PR DESCRIPTION
**Pull request by RamilGaripov**
Ticket: #metriport/metriport-internal/issues/2510

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/3060

### Description
- Added a ton more improvements to the Bundle slimming algorithm: 
  - Removing certain DiagnosticReports from inputs based on their category 
  - Only keeping the most recent 3 DiagnosticReports per doctor, per category
- Added token usage, cost calculations, and analytics

### Testing

- Local
  - [x] Test with a handful of Bundles of different sizes
- Staging
  - [ ] Test with a couple of staging patient Bundles
- Production
  - [ ] Monitor analytics and costs

### Release Plan
- [ ] Merge this

### CODE REVIEW

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 1]
```diff
@@ -52,117 +59,205 @@ const referenceResources = [
...
-  const dateFrom = dayjs().subtract(numHistoricalYears, "year").format(ISO_DATE);
+  log(`Starting with requestId ${requestId}, and bundle length ${bundle.entry?.length}`);
+  const latestReportDate = findDiagnosticReportResources(bundle)
+    .flatMap(report => {
+      return getDatesFromEffectiveDateTimeOrPeriod(report);
+    })
+    .filter((date): date is string => date !== undefined)
+    .sort((a, b) => b.localeCompare(a))[0];
+  const numHistoricalYears = 1;
+  const initialDate = dayjs(latestReportDate) ?? dayjs();
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:125]: let's use the `buildDayJs` function please

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 2]
```diff
@@ -52,117 +59,205 @@ const referenceResources = [
...
-  const numHistoricalYears = 2;
-  const dateFrom = dayjs().subtract(numHistoricalYears, "year").format(ISO_DATE);
+  log(`Starting with requestId ${requestId}, and bundle length ${bundle.entry?.length}`);
+  const latestReportDate = findDiagnosticReportResources(bundle)
+    .flatMap(report => {
+      return getDatesFromEffectiveDateTimeOrPeriod(report);
+    })
+    .filter((date): date is string => date !== undefined)
+    .sort((a, b) => b.localeCompare(a))[0];
+  const numHistoricalYears = 1;
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:124]: Shouldn't this be on the top of the file?

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 3]
```diff
@@ -326,11 +468,11 @@ function applyResourceSpecificFilters(res: Resource): any | undefined {
     updRes.status = Array.from(
       new Set(res.clinicalStatus?.coding?.flatMap(coding => coding.code || []))
     ).join(", ");
-    if (updRes.status === "") delete updRes.status;
+    if (isUselessStatus(updRes.status)) delete updRes.status;
   }
 
   if (res.resourceType === "Immunization") {
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:474]: code after the cloneDeep should use `updRes`, right?

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 4]
```diff
@@ -341,67 +483,114 @@ function applyResourceSpecificFilters(res: Resource): any | undefined {
...
+    if (allTypes) {
+      let removeResource = false;
+      allTypes.forEach(type => {
+        if (
+          REPORT_TYPES_BLACKLIST.includes(type) ||
+          REPORT_TYPES_BLACKLIST.some(blacklistedType => type.includes(blacklistedType))
+        )
+          removeResource = true;
+      });
+      if (removeResource) return undefined;
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:537]: nit: could just use a regular loop and return if the condition is met, no need for variable and additional `if`

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 5]
```diff
@@ -605,53 +814,62 @@ function replaceReferencesWithData(
...
+
+        performers.forEach(p => {
+          const ref = p.reference;
+          if (!ref) return;
+
+          referencedIds.add(ref);
+          const performer = map.get(ref);
+
+          if (performer?.resourceType === "Organization") {
+            const name = typeof performer.name === "string" ? performer.name : "";
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:835]: .trim() on name

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 6]
```diff
@@ -605,53 +814,62 @@ function replaceReferencesWithData(
...
+            if (name.length > 0) orgs.push(name);
+          } else if (performer?.resourceType === "Practitioner") {
+            const name =
+              typeof performer.name === "string" ? performer.name : getNameString(performer.name);
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:839]: ditto: .trim() on .name

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 7]
```diff
@@ -605,53 +814,62 @@ function replaceReferencesWithData(
...
+            if (name) practitioners.push(name);
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:840]: shouldn't check if length > 0, to match the logic on the org above?

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 8]
```diff
@@ -731,7 +949,135 @@ function replaceReferencesWithData(
...
+          if (formDataSet.has(sentence)) {
+            return false;
+          }
+          formDataSet.add(sentence);
+          return true;
+        });
+        newPresentedForm.push(filteredSentences.join(". "));
+      });
+
+      const filsss = newPresentedForm.join("\n");
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:1074]: prob a typo on `filsss`?

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 9]
```diff
@@ -744,3 +1090,19 @@ function getNameString(names: HumanName | HumanName[] | undefined): string | und
...
+  return status === "" || status === "final";
+}
+
+function calculateCostsBasedOnTokens(totalTokens: { input: number; output: number }): {
+  input: number;
+  output: number;
+  total: number;
+} {
+  const input = (totalTokens.input / 1000) * 0.0015;
+  const output = (totalTokens.output / 1000) * 0.0075;
```
Comment (leite08) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:1104]: What are those numbers? Let's create consts for them on the top of the file?

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 1 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:125]: Fixed!

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 2 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:124]: yea, good point

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 3 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:474]: In this case, checking the type of the initial resource allows us to keep track of types of properties inside it. If I do
`  if (updRes.resourceType === "Immunization") {`

then the types attributes after will not be known

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 4 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:537]: yup, 💯

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 5 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:835]: added

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 6 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:839]: actually, this check if it's a string is a left-over. Practitioner.name is never a string.

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 8 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:1074]: oh crap, yes.. sometimes i use random strings for var names when im not sure what im gonna name them.. but this time i forgot to update it 😅

**Review by leite08 — APPROVED** [review]

**Pull request merged**